### PR TITLE
Fix NullpointerException trigger middlemanager unusable when using httpRemote 

### PR DESCRIPTION
**Pull request by liuxiaohui1221**
fix NullpointerException trigger middlemanager unusable when using `httpRemote`

Fixes #XXXX.

### Description

Task of HttpRemoteTaskRunnerWorkItem null will make Overlord trigger NullpointerException when found middlemanager node registered, lead to the failure to register, the middlemanager follow-up also won't provide services.  
When the Overlord receives the task change response from MiddleManager, it discovers that the tasks in HttpRemoteTaskRunner do not have a corresponding task.  But the task is in the taskStorage would trigger to create a task to null HttpRemoteTaskRunnerWorkItem,  
If so, the MiddleManager node will trigger a null call to task due to re-registration.  

Overlord log:

```
2021-12-15T14:08:06,539 ERROR [CuratorDruidNodeDiscoveryProvider-ListenerExecutor] org.apache.druid.discovery.DruidNodeDiscoveryProvider$ServiceDruidNodeDiscovery - Listener[org.apache.druid.indexing.overlord.hrtr.HttpRemoteTaskRunner$1@59636370].nodesAdded([DiscoveryDruidNode{druidNode=DruidNode{serviceName='druid/middleManager', host='br-zeus-private-stress02', bindOnHost=false, port=-1, plaintextPort=8091, enablePlaintextPort=true, tlsPort=-1, enableTlsPort=false}, nodeRole='MIDDLE_MANAGER', services={workerNodeService=WorkerNodeService{ip='br-zeus-private-stress02', capacity=9, version='0', category='_default_worker_category'}}}]) threw exception. Ignored.
java.lang.NullPointerException: null
	at org.apache.druid.indexing.worker.TaskAnnouncement.create(TaskAnnouncement.java:48) ~[druid-indexing-service-0.20.1.4.jar:0.20.1.4]
	at org.apache.druid.indexing.overlord.hrtr.HttpRemoteTaskRunner.addWorker(HttpRemoteTaskRunner.java:537) ~[druid-indexing-service-0.20.1.4.jar:0.20.1.4]
	at org.apache.druid.indexing.overlord.hrtr.HttpRemoteTaskRunner.access$200(HttpRemoteTaskRunner.java:122) ~[druid-indexing-service-0.20.1.4.jar:0.20.1.4]
	at org.apache.druid.indexing.overlord.hrtr.HttpRemoteTaskRunner$1.lambda$nodesAdded$0(HttpRemoteTaskRunner.java:470) ~[druid-indexing-service-0.20.1.4.jar:0.20.1.4]
	at java.util.ArrayList.forEach(ArrayList.java:1257) ~[?:1.8.0_171]
	at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1080) ~[?:1.8.0_171]
	at org.apache.druid.indexing.overlord.hrtr.HttpRemoteTaskRunner$1.nodesAdded(HttpRemoteTaskRunner.java:470) ~[druid-indexing-service-0.20.1.4.jar:0.20.1.4]
	at org.apache.druid.discovery.DruidNodeDiscoveryProvider$ServiceDruidNodeDiscovery$FilteringUpstreamListener.nodesAdded(DruidNodeDiscoveryProvider.java:166) ~[druid-server-0.20.1.4.jar:0.20.1.4]
	at org.apache.druid.curator.discovery.CuratorDruidNodeDiscoveryProvider$NodeRoleWatcher.lambda$addNode$2(CuratorDruidNodeDiscoveryProvider.java:339) ~[druid-server-0.20.1.4.jar:0.20.1.4]
	at org.apache.druid.curator.discovery.CuratorDruidNodeDiscoveryProvider$NodeRoleWatcher.lambda$safeSchedule$5(CuratorDruidNodeDiscoveryProvider.java:458) ~[druid-server-0.20.1.4.jar:0.20.1.4]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_171]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_171]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_171]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_171]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_171]
```

Fix: When middlemanager register, filtering processing of Task of HttpRemoteTaskRunnerWorkItem null 

<hr>

##### Key changed/added classes in this PR

- `org.apache.druid.indexing.overlord.hrtr.HttpRemoteTaskRunner`

<hr>

This PR has:

- [x] been self-reviewed.
      - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by FrankChen021** [comment]
Thanks for your first contribution. Is it possible to add a UT case?

**Comment by liuxiaohui1221** [comment]
> Thanks for your first contribution. Is it possible to add a UT case?

OK,I will try to add it  when I have time.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/hrtr/HttpRemoteTaskRunner.java [hunk 1]
```diff
@@ -583,7 +583,7 @@ private void addWorker(final Worker worker)
           for (Map.Entry<String, HttpRemoteTaskRunnerWorkItem> e : tasks.entrySet()) {
             if (e.getValue().getState() == HttpRemoteTaskRunnerWorkItem.State.RUNNING) {
               Worker w = e.getValue().getWorker();
-              if (w != null && w.getHost().equals(worker.getHost())) {
+              if (w != null && w.getHost().equals(worker.getHost()) && e.getValue().getTask() != null) {
```
Comment (suneet-s) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/hrtr/HttpRemoteTaskRunner.java:586]: I do not know this part of the system well, but I found a couple other places where this is called without a null check - `getPendingTaskPayloads` and `runTaskOnWorker`.

Could you annotate `HttpRemoteTaskRunner#getTask` with `@Nullable` so we can use intelliJ inspections to catch any uses of `getTask` without a null check.

**Comment by tilakchowdary** [comment]
We are facing the same issue

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/hrtr/HttpRemoteTaskRunner.java (hunk 1 above)
Comment (wiquan) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/hrtr/HttpRemoteTaskRunner.java:586]: I see no risk in what was added, and your comment just delays the fix getting into the upstream. There is no mention or requirement for IntelliJ coding standards/annotations or whatever. If that bothers you that its not ideal in your IDE, why don't you raise a PR for IntelliJ (<your IDE here>) annotations and cleanups. Most likely the OP has already fixed in his fork and made this PR for upstream and the rest of us.  Now we have a stale PR (a valid fix) for something others need. Please get real, there are over 1300 open issues. Momentum would be nice, right?

Comment (abhishekagarwal87) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/hrtr/HttpRemoteTaskRunner.java:586]: PR Authors can always agree or disagree with the code review. A UT was also requested that has not been added yet. While we want to fix the open issues, we do not want to compromise on the code quality. 

Also, IntelliJ inspections is a test job that we run for each PR. Suneet is referring to the inspections that run in the test job, not a local IDE.

**Review by gianm — APPROVED** [review]
The fix LGTM. Apologies that this one fell through the cracks. I'm going to merge it, since it doesn't seem like people have had the time to add unit testing or annotations to enable static analysis, but we still want to get the bug patched.

**Comment by gianm** [comment]
thanks @liuxiaohui1221!

**Pull request merged**